cg/services/order_validation_service/order_validation_service.py

```diff
@@ -6,3 +8,9 @@ class OrderValidationService:
 
     def __init__(self, tomte_service: TomteOrderValidationService):
         self.tomte_service = tomte_service
+
+    def validate(self, order_json: str):
+        order = Order.model_validate_json(order_json)
+
+        if order.workflow == Workflow.TOMTE:
+            self.tomte_service.validate(order_json)
```

Comment:
Are just going to add more ifs here? Is a dict not better suited?
```python
def __init__(self, some_params...)
    self.workflow_validators: dict[Workflow, Any] = {
        Workflow.TOMTE = TomteOrderValidationService(some_params...)
    }

def validate(self, order_json: str):
    self.workflow_validators[order.workflow](order_json)
```

Reply:
I think I prefer explicitly passing each of the services in the constructor. There is probably a better design, but I don't think using a dict is better.

Reply:
Yes, I prefer passing the services explicitly in the constructor. We can keep the services in a dict internally, but I'll leave it for now. We don't want to have complex logic for instantiating the validation services here.

There might be a better design than an "orchestrator" validation service, but it is the best I could think of for now.

Reply:
I see the point of extracting the instantiation. But passing each service as an argument will grow to be quite ugly with the 14 validation_services we have planned. Would it make sense to pass a dict to this service containing instantiated objects?

Reply:
There is something called service locator pattern which might be a better fit for us.

Basically, we would let all validation services implement an `OrderValidationService` interface and instantiate them centrally in some container.

Then we would retrieve them like `validation_service = service_locator.get_service(order.workflow)`.

What do you think? It would look something like this

```python

def validate_order(order_json: str):
    order = Order.model_validate_json(order_json)
    validation_service: OrderValidationService = service_locator.get_service(order.workflow)
    validation_service.validate(order)
```

Reply:
I think what you want to do is along the lines of:

```suggestion
class OrderValidator:
    def __init__(self, validation_service: ValidationService = TomteValidationService()):
        self.validation_service = validation_service

    def validate(self, order_json: str):
            self.tomte_service.validate(order_json)
```

Then the place where you would need to validate something:

```
 # create dependency
validationService = TomteOrderValidationService()
# inject dependency
validator = OrderValidator(validationService)
validator.validate(order_json)
```
Then you circumvent the if statements that will need to grow with each worfklow, right?

Reply:
I was curious about the "registry" and "service locator" and looked it up a bit. Is this really what we want? It seems like it is akin to what we wanted to get away from by not passing the CgConfig directly, instead being more explicit by passing e.g. the store / housekeeper api. 

Some things I found about using a service locator/ registry / context also:
https://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/
https://blog.ploeh.dk/2014/05/15/service-locator-violates-solid/
https://blog.ploeh.dk/2015/10/26/service-locator-violates-encapsulation/

this video was quite ifnoramtive:
https://www.youtube.com/watch?v=RlfLCWKxHJ0

I can see what they are talking about given our test suite where we for example have to pass the whole config or just create a context. It is often quite painful. 

Maybe I am missing something?

Reply:
A registry is the first thing ChatGPT suggests when I describe what we want to do and to me it makes sense if we map an order type to its appropriate validator. In his examples the Services seem a bit more diverse than what we have, like roof, house, window. While we have chefs knife, butcher knife, gutting knife etc.

I think what is important for us is that we can extend it easily which is something they don't give a solution for 🤔

Reply:
But I feel like all these examples dismantled our CGContext pretty well...

Reply:
The problem with `CgConfig` is that we pass it down to classes, that is a bad design. It should be used at a top level, in a CLI command or endpoint, retrieving a service and using it.

I think service locator or a factory pattern will work well for our order validation services

Reply:
I think what we do not want to do is pass a service locator in the constructor of the service. 
But rather get the "right" service first and then inject the service and not the service locator. If you want to do it like that.

Reply:
Alternatively, we can setup a dependency injection framework (similar to what we have in trailblazer) and let it automatically resolve the dependencies 😄 I guess our validation logic is an edge case, like Vincent said it is a bit different because we have many flavors of the same thing. I'll think about it some more today

Reply:
This is what you get with my suggestion above if I understand the strategy approach correctly.

Each service (tomte, rare-disease, mip, fluffy and whatnot) will have a OrderValidationService that you inject to the validator that you then call in the needed places. 

I think the question here is "with what strategy do we retrieve the right service". Honeslty a workflow to service map might be the easiest to expand, maintain and understand.

Reply:
I agree, implementing a draft now 👍